src/express-lane-auction/ExpressLaneAuction.sol

```diff
@@ -0,0 +1,477 @@
+    function setReservePrice(uint256 newReservePrice) external onlyRole(RESERVE_SETTER_ROLE) {
+        if (isReserveBlackout()) {
+            revert ReserveBlackout();
+        }
+
+        _setReservePrice(newReservePrice);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function balanceOf(address account) public view returns (uint256) {
+        return _balanceOf[account].balanceAtRound(currentRound());
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function withdrawableBalance(address account) public view returns (uint256) {
+        // CHRIS: TODO: consider whether the whole balance of mapping and the round number should be in a lib together
+        return _balanceOf[account].withdrawableBalanceAtRound(currentRound());
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function deposit(uint256 amount) external {
+        _balanceOf[msg.sender].increase(amount);
+        biddingToken.safeTransferFrom(msg.sender, address(this), amount);
+        emit Deposit(msg.sender, amount);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function initiateWithdrawal() external {
+        uint64 curRnd = currentRound();
+        uint256 amount = _balanceOf[msg.sender].balance;
+        _balanceOf[msg.sender].initiateWithdrawal(curRnd);
+        // CHRIS: TODO: we have + 2 here, that's leaking an implementation detail
+        emit WithdrawalInitiated(msg.sender, amount, curRnd + 2);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function finalizeWithdrawal() external {
+        uint256 amountReduced = _balanceOf[msg.sender].finalizeWithdrawal(currentRound());
+        biddingToken.safeTransfer(msg.sender, amountReduced);
+        // CHRIS: TODO: consider adding the following assertion - it's an invariant
+        // CHRIS: TODO: Invariant: assert(withdrawableBalance(msg.sender) == 0);
+        emit WithdrawalFinalized(msg.sender, amountReduced);
+    }
+
+    /// @dev Update local state to resolve an auction
+    /// @param isMultiBid Where the auction should be resolved from multiple bids
+    /// @param firstPriceBid The winning bid
+    /// @param firstPriceBidder The winning bidder
+    /// @param priceToPay The price that needs to be paid by the winner
+    /// @param biddingInRound The round bidding is taking place in. This is not the round the bidding is taking place for, which is biddingInRound + 1
+    function resolveAuction(
+        bool isMultiBid,
+        Bid calldata firstPriceBid,
+        address firstPriceBidder,
+        uint256 priceToPay,
+        uint64 biddingInRound
+    ) internal {
+        // store that a round has been resolved
+        uint64 biddingForRound = biddingInRound + 1;
+        latestResolvedRounds.setResolvedRound(biddingForRound, firstPriceBid.expressLaneController);
+
+        // first price bidder pays the beneficiary
+        _balanceOf[firstPriceBidder].reduce(priceToPay, biddingInRound);
+        biddingToken.safeTransfer(beneficiary, priceToPay);
```

Comment:
if we want to make it more efficient, we can store the balance change without transferring everytime

Reply:
that might be good, but I cant really think of a nice way to do it. I dont really want to add a withdraw mechanism since we dont know if the beneficiary will be able to call that, and tracking when we should transfer otherwise is a bit difficult. We could try to transfer on round % 10 or something, but we cant be sure that the auction will be resolved on those rounds, the auctioneer would need to be trusted to eventually call it on a round % 10.

Reply:
>  if the beneficiary will be able to call that
It doesn't need to be onlyBeneficiary, it only need to be a separate call

Reply:
kk, will add this as a TODO for potential optimization